src/Service/Image.php

```diff
@@ -0,0 +1,122 @@
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Filesystem\Path;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+final class Image implements ImageInterface
+{
+    public function __construct(
+        private readonly HttpClientInterface $client,
+        private readonly string $publicPath,
+        private readonly MessageBusInterface $messageBus,
+    ) {
+    }
+
+    /**
+     * @throws FilesystemException
+     * @throws TransportExceptionInterface
+     * @throws ImageFetchException
+     */
+    public function fetch(string $url): string
+    {
+        $filesystem = new Filesystem();
+
+        try {
+            $path = $this->generatePath(url: $url, absolute: true);
+            $filesystem->mkdir(
+                Path::normalize($path),
+                0775
+            );
+        } catch (IOExceptionInterface $exception) {
+            throw new FilesystemException('Unable to create upload folder', $exception->getCode(), $exception);
+        }
+
+        $response = $this->client->request('GET', $url);
+        if (200 !== $response->getStatusCode()) {
+            throw new ImageFetchException(sprintf('Failed to fetch %s with code %s', $url, $response->getStatusCode()), $response->getStatusCode());
+        }
+
+        $dest = $path.basename($url);
+        $fileHandler = fopen($dest, 'w');
+        foreach ($this->client->stream($response) as $chunk) {
+            fwrite($fileHandler, $chunk->getContent());
+        }
+        fclose($fileHandler);
+
+        return $this->getRelativePath($dest);
+    }
+
+    public function remove(\App\Entity\Image $image): bool
+    {
+        // TODO: Implement remove() method.
+        return false;
+    }
+
+    public function transform(\App\Entity\Image $image): void
+    {
+        $path = $image->getLocal();
+        if (!is_null($path)) {
+            $this->messageBus->dispatch(new WarmupCache($this->getRelativePath($path)));
+        }
+    }
+
+    /**
+     * Get relative path from absolute path.
+     *
+     * @param string $path
+     *   The absolute path
+     *
+     * @return string
+     *   Path stripped the image web-root
+     */
+    private function getRelativePath(string $path): string
+    {
+        return preg_replace('/^'.preg_quote($this->publicPath, '/').'/', '', $path);
+    }
+
+    /**
+     * Generate safe path to storage file in based on URL.
+     *
+     * @param string $url
+     *   The files URL
+     * @param int<1, max> $depth
+     *   The depth of the generated path
+     * @param int<1, max> $size
+     *   The size of each element in the generated path
+     *
+     * @return string
+     *   The generated path ending with slash
+     */
+    private function generatePath(string $url, bool $absolute = false, int $depth = 1, int $size = 8): string
```

Comment:
Why do we want to generate a path from an image URL. Is it better for the filesystem to put files in different folders?

I would just hash the full URL and save all the images in a single folder (cf. comment above).